buildRustPackage: allow passing in cargoDeps

SuperSandro2000 opened the pull request:
## Description of changes

Since roughly a week or so the following overlay fails and I couldn't so far figure out why because the changes done in the last ~3 months or so look rather innocent.

```
  # TODO: upstream into nixpkgs
  wezterm-headless = rustPlatform.buildRustPackage {
    pname = "${wezterm.pname}-headless";
    inherit (wezterm) version src postPatch;
    inherit (wezterm.passthru) cargoLock;

    nativeBuildInputs = [ pkg-config ];

    buildInputs = [ openssl ];

    cargoBuildFlags = [
      "--package" "wezterm"
      "--package" "wezterm-mux-server"
    ];

    doCheck = false;

    postInstall = /* bash */ ''
      install -Dm644 assets/shell-integration/wezterm.sh -t $out/etc/profile.d
      sed -e 's|hash wezterm 2>/dev/null|command type -P wezterm &>/dev/null|g' \
        -e 's|hash base64 2>/dev/null|command type -P base64 &>/dev/null|g' \
        -e 's|hash hostname 2>/dev/null|command type -P hostname &>/dev/null|g' \
        -e 's|hash hostnamectl 2>/dev/null|command type -P hostnamectl &>/dev/null|g' \
        -i $out/etc/profile.d/wezterm.sh

      install -Dm644 ${wezterm.terminfo}/share/terminfo/w/wezterm -t $out/share/terminfo/w
    '';
  };
```

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

winterqt reviewed (changes requested):
The overlay you show has never worked; we've never exposed `cargoLock` in passthru, only `cargoDeps`. Are you sure you don't mean to use `wezterm.cargoDeps` (or `wezterm.passthru.cargoDeps`) in it?

SuperSandro2000 commented:
Derp 🤦🏼 I found the commit which I dropped by accident while rebasing https://github.com/NixOS/nixpkgs/commit/4f4c193117233e103c627dcb42cae9a0a278599c

> Are you sure you don't mean to use wezterm.cargoDeps (or wezterm.passthru.cargoDeps) in it?

In essence I want this but it fails on https://github.com/nixos/nixpkgs/blob/bcbfad30628c60e310e5f7c95436de2f0c05921e/pkgs/build-support/rust/build-rust-package/default.nix#L62-L64

Maybe that should be extended instead?

SuperSandro2000 committed:
buildRustPackage: allow passing in cargoDeps

This is useful when constructing overlays and trying to copy a
buildRustPackage from a different package

SuperSandro2000 commented:
@winterqt ping

TomaSajt commented:
Probably superseded by https://github.com/NixOS/nixpkgs/pull/354999
